assets/js/interactivity/cart/cart-store.ts

```diff
@@ -0,0 +1,25 @@
+/**
+ * External dependencies
+ */
+import { select, subscribe } from '@wordpress/data';
+import { CART_STORE_KEY } from '@woocommerce/block-data';
+import { updateStore } from '../store';
+
+export const syncStoreWithInteractivityState = () => {
+	const store = select( CART_STORE_KEY );
+	const cartData = store.getCartData();
+	const isResolutionFinished =
+		select( CART_STORE_KEY ).hasFinishedResolution( 'getCartData' );
+
+	if ( isResolutionFinished ) {
+		updateStore( {
+			state: {
+				woocommerce: {
+					cart: cartData,
+				},
+			},
+		} );
```

Comment:
You don't really need a special function to update the store. You can do:

```js
state.woocommerce.cart = cartData;
```

Also, this logic should not be inside the Interactivity API per se. Take into account that this package will disappear once the Interactivity API is merged with WordPress Core. 

Ideally, we should do it within a `data-wp-init` callback, which is a `useEffect` with access to state. The problem is that we should add it only once per page, and we don't have any pattern for that. We need to think about this.

Reply:
I opened a [discussion](https://github.com/WordPress/gutenberg/discussions/52514) to discuss ideas about this.

Reply:
> Also, this logic should not be inside the Interactivity API per se. Take into account that this package will disappear once the Interactivity API is merged with WordPress Core.

I agree, but I had some issues with this logic outside the interactivity API package. Since it is a temporary thing (since the Interactivity API will be included in Core) and since I created a dedicated folder, we could keep this approach, for now. What do you think?

Reply:
Yes, we can merge it as it is and change it later once we figure out the right pattern.

Reply:
Let's try to implement [David's idea](https://github.com/WordPress/gutenberg/discussions/52514#discussioncomment-6419346) of a special callback passed to `store`:

```js
// product-button-view.js
import { store } from "@woocommerce/interactivity";

store(
  {
    // Regular store...
  },
  {
    afterLoad: ({ state }) => {
      const reduxStore = getReduxStore();
      reduxStore.subscribe(() => {
        state.woocommerce.cart = reduxStore.getState();
      });
      dispatch(reduxStore.fetchStoreAPI()); // Fetch the store.
    },
  }
);
```

If you are concerned about the initial load performance, you can wrap the `dispatch` with a `requestIdleCallback`.

If multiple blocks need to use the same callback, you could move it to a different file. That way, it will only load once even if there are multiple blocks on the page.

```js
// add-to-cart-button-view.js
import { store } from "@woocommerce/interactivity";
import "../redux-sync.js";

store({
  // Local store needs...
});
```

```js
// add-to-cart-view.js
import { store } from "@woocommerce/interactivity";
import "../redux-sync.js";

store({
  // Local store needs...
});
```

```js
// redux-sycn.js
import { store } from "@woocommerce/interactivity";

store(
  {
    // Your global store needs...
  },
  {
    afterLoad: ({ state }) => {
      const reduxStore = getReduxStore();
      reduxStore.subscribe(() => {
        state.woocommerce.cart = reduxStore.getState();
      });
      dispatch(reduxStore.fetchStoreAPI()); // Fetch the store.
    },
  }
);
```



@gigitux: Let us know what you think! If you like the idea, we'll add it to our tasks and we'll take care of the implementation.

cc: @darerodz

Reply:
I think that it makes sense. Do you think that is there a way to init the redux store only when the page is cached? 🤔

Reply:
> Do you think that is there a way to init the redux store only when the page is cached?

Yeah, that should be possible. You just have to add that information to the state:

```php
if ( isset( WC()->cart ) ) {
  // Add user info...
  wp_store([
    'state' => [
      'woocommerce' => [
        'syncStore' => false
      ]
    ]
  ]);
} else {
  // Don't add user info...
  wp_store([
    'state' => [
      'woocommerce' => [
        'syncStore' => true
      ]
    ]
  ]);
}
```

Then, use it in the `afterLoad` callback:

```js
store(
  {},
  {
    afterLoad: ({ state }) => {
      if (state.woocommerce.syncStore) {
        // Sync the store...
      }
    },
  }
);
```

Reply:
I'd love to implement the `afterLoad` callback. 😄

Is it required to include it here? Or could I start working on it in a different PR?

Reply:
Thanks, David. I forgot to tell you yesterday 😄

Maybe different PR and then rebase/merge in this one to keep things organized?

Reply:
I've just opened a PR with a preliminary implementation for the Interactivity Store callbacks.
- https://github.com/woocommerce/woocommerce-blocks/pull/10338

Reply:
> If multiple blocks need to use the same callback, you could move it to a different file. That way, it will only load once even if there are multiple blocks on the page.

@luisherranz, just pointing out that we'd have to pay attention to the Webpack bundling so we are not duplicating the same callback. ☝️